### ISSUE TITLE
Expose lint levels from CLI

### ISSUE THREAD

**Issue by Kixunil** [rust-lang/cargo#14522]
### Problem

There are cases when it's not desirable configuring the lint levels using `Cargo.toml` and `RUST(DOC)FLAGS` is not discoverable and also has the footgun that `RUSTFLAGS` != `RUSTDOCFLAGS`. Specifically it's useful to have things like `#![warn(missing_docs)]` in the `lib.rs` but deny the warning in the CI because it's annoying to have to document everything if one is just designing the API and manually editing the file (potentially accidentally committing it) is also annoying. Editing `Cargo.toml` from CI's shell script is fragile.

### Proposed Solution

It'd be best to just expose the lint levels from `cargo` and make sure `cargo` sets them correctly for rustc and rustdoc. Also please don't forget forbid because that level is also useful and even if it's not super-important it'll be probably just a few more lines once other lint levels are in.

### Notes

I've opened this as requested in https://github.com/rust-lang/cargo/issues/12739#issuecomment-2338147752

**Comment by epage** [comment]
> Specifically it's useful to have things like #![warn(missing_docs)] in the lib.rs but deny the warning in the CI because it's annoying to have to document everything if one is just designing the API and manually editing the file (potentially accidentally committing it) is also annoying. 

Is #8424 sufficient for this use case?

**Comment by Kixunil** [comment]
For that specific use case yes, it is. But there is also at least one different use case.

**Comment by epage** [comment]
Could you elaborate on that?  I'm not seeing it mentioned in the issue and we'd need to understand requirements to be able to discuss designs.

**Comment by Kixunil** [comment]
Basically I need `-F unsafe-code` to check is a crate contains `unsafe` because `unsafe` code needs (expensive) `miri` test but safe doesn't (and this is nuanced which is why I don't think `miri` should be doing this decision). `--forbid-unsafe` would do the job but it looks really ad-hoc "why only forbid?" and "why only unsafe?".

**Comment by epage** [comment]
For me, the question is if `-F unsafe-code` has enough of a use case to justify a specific or general flag over still pointing to `RUSTFLAGS` / `RUSTDOCFLAGS`.

**Comment by Kixunil** [comment]
Maybe it's other way around. If there's use case for `--allow`/`--warn`/`--deny` then it'd be weird not to add `--forbid`. It's hard to believe that nobody would need those but maybe it is the case.

**Comment by epage** [comment]
> Maybe it's other way around. If there's use case for --allow/--warn/--deny then it'd be weird not to add --forbid. It's hard to believe that nobody would need those but maybe it is the case.

Yes, we would handle all of those together.  The point I was raising was whether there is enough of a use case for those general flags to exist in Cargo's CLI.

**Comment by epage** [comment]
I wonder if that case would be another use case for a new lint level which I proposed at https://internals.rust-lang.org/t/forbid-deny-warn-allow-and-notice/19986

**Comment by Kixunil** [comment]
Maybe if you really, really don't want to pollute CLI, a different way to solve these issues would be to have something like `--override-cargo-toml override.toml` option, the override file could contain anything `Cargo.toml` and conflicting values would get replaced by the override. That way we could override settings in `Cargo.toml` without it being fragile. One other use case that comes to mind is to have `publish = false` by default and using the override when publishing.

**Comment by epage** [comment]
> One other use case that comes to mind is to have publish = false by default and using the override when publishing.

For myself, that seems strange for publish to change like that and calls out a good complication to having an "override": it complicates provenance for that crate itself.
